src/FSharp Examples/AdversarialExampleGeneration.fs

```diff
@@ -0,0 +1,134 @@
+module TorchSharp.Examples.AdversarialExampleGeneration
+
+open System
+open System.IO
+open System.Diagnostics
+
+open TorchSharp
+open TorchSharp.Tensor
+open TorchSharp.NN
+
+open type TorchSharp.NN.Modules
+open type TorchSharp.TorchScalar
+
+open TorchSharp.Examples
+
+/// <summary>
+/// FGSM Attack
+///
+/// Based on : https://pytorch.org/tutorials/beginner/fgsm_tutorial.html
+/// </summary>
+/// <remarks>
+/// There are at least two interesting data sets to use with this example:
+/// 
+/// 1. The classic MNIST set of 60000 images of handwritten digits.
+///
+///     It is available at: http://yann.lecun.com/exdb/mnist/
+///     
+/// 2. The 'fashion-mnist' data set, which has the exact same file names and format as MNIST, but is a harder
+///    data set to train on. It's just as large as MNIST, and has the same 60/10 split of training and test
+///    data.
+///    It is available at: https://github.com/zalandoresearch/fashion-mnist/tree/master/data/fashion
+///
+/// In each case, there are four .gz files to download. Place them in a folder and then point the '_dataLocation'
+/// constant below at the folder location.
+///
+/// The example is based on the PyTorch tutorial, but the results from attacking the model are very different from
+/// what the tutorial article notes, at least on the machine where it was developed. There is an order-of-magnitude lower
+/// drop-off in accuracy in this version. That said, when running the PyTorch tutorial on the same machine, the
+/// accuracy trajectories are the same between .NET and Python. If the base convulutational model is trained
+/// using Python, and then used for the FGSM attack in both .NET and Python, the drop-off trajectories are extremenly
+/// close.
+/// </remarks>
+
+let mutable trainBatchSize = 64
+let mutable testBatchSize = 128
+
+let logInterval = 100
+
+let cmdArgs = Environment.GetCommandLineArgs()
+let dataset = if cmdArgs.Length = 2 then cmdArgs.[1] else "mnist"
+
+let datasetPath = Path.Join(Environment.GetFolderPath(Environment.SpecialFolder.DesktopDirectory), "..", "Downloads", dataset)
+
+Torch.SetSeed(1L)
+
+let hasCUDA = Torch.IsCudaAvailable()
+
+let device = if hasCUDA then Device.CUDA else Device.CPU
+
+let criterion x y = Functions.nll_loss().Invoke(x,y)
+
+let attack (image:TorchTensor) (eps:TorchScalar) (data_grad:TorchTensor) =
+    use sign = data_grad.sign()
+    (image + eps * sign).clamp(0.0.ToScalar(), 1.0.ToScalar())
+
+let test (model:MNIST.Model) (eps:float) (dataLoader:MNISTReader) size =
+
+    let mutable correct = 0
+
+    for (input,labels) in dataLoader do
+
+        input.requires_grad <- true
+        
+        begin  // This is introduced in order to let a few tensors go out of scope before GC
```

Comment:
Is this idiomatic F#?  
This is not a complaint ... I was facing a similar situation and was wondering what's an optimal syntax in such case.  
Would a `do` block work here as well?  
Maybe a question for @cartermp?

Reply:
It's not particularly idomatic but I think nothing is - it's a rare situation.  A 'do' would be ok as well, as would parentheses

Reply:
Yes, the need to clean up GPU memory ASAP does introduce some non-attractive code here and there. Same thing in C#, with lots of 'usings' that are there for no other reason.

Reply:
Another alternative could be the `using` function:

```fsharp
using (input --> model) (fun estimate ->
	use loss = criterion estimate labels

	model.ZeroGrad()
	loss.backward()

	use perturned = attack input (eps.ToScalar()) (input.grad())
	use final = perturbed --> model
	correct <- correct + final.argmax(1L).eq(labels).sum().ToInt32()
)
```

I believe that this can achieve the same thing without an explicitly scoped block. The downside of course is that we're mixing two kinds of "bind a disposable resource" approaches.

Reply:
This seems like a step in the wrong direction, IMO -- greater syntactic overhead.